# 评价类模型真题分享讨论
总结人：吴欣怡、张闻韬

## 2021年C题 生产企业原材料的订购与运输
主讲人：庄树格

某建筑和装饰板材的生产企业所用原材料主要是木质纤维和其他植物素纤维材料,总体可分为 A，B，C 三种类型。该企业每年按 48 周安排生产，需要提前制定 24 周的原材料订购和转运计划，即根据产能要求确定需要订购的原材料供应商（称为“供应商”）和相应每周的原材料订购数量（称为“订货量”），确定第三方物流公司（称为“转运商”）并委托其将供应商每周的原材料供货数量（称为“供货量”）转运到企业仓库。
该企业每周的产能为 2.82 万立方米，每立方米产品需消耗 A 类原材料 0.6 立方米，或 B 类原材料 0.66 立方米，或 C 类原材料 0.72 立方米。由于原材料的特殊性，供应商不能保证严格按订货量供货，实际供货量可能多于或少于订货量。为了保证正常生产的
需要，该企业要尽可能保持不少于满足两周生产需求的原材料库存量，为此该企业对供应商实际提供的原材料总是全部收购。在实际转运过程中，原材料会有一定的损耗（损耗量占供货量的百分比称为“损耗率”），转运商实际运送到企业仓库的原材料数量称为“接收量”。每家转运商的运输
能力为 6000 立方米/周。通常情况下，一家供应商每周供应的原材料尽量由一家转运商运输。原材料的采购成本直接影响到企业的生产效益，实际中 A 类和 B 类原材料的采购单价分别比 C 类原材料高 20%和 10%。三类原材料运输和储存的单位费用相同。
附件 1 给出了该企业近 5 年 402 家原材料供应商的订货量和供货量数据。附件 2 给出了 8 家转运商的运输损耗率数据。请你们团队结合实际情况，对相关数据进行深入分析，研究下列问题：
### 问题1
根据附件 1，对 402 家供应商的供货特征进行量化分析，建立反映保障企业生产重要性的数学模型，在此基础上确定 50 家最重要的供应商，并在论文中列表给出结果。

#### 附件1（企业的订货量$(m^3)$）
![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig1.png)
#### 附件2 （供应商的供货量$(m^3)$）
![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig2.png)
### 建模部分
#### 问题分析
该问主要对 402 家供货商的情况进行整体评价分析，以便于后续方案的建模与求解。本题中假设A，B，C三种材料等价，则仅从订货量和供给情况出发。评价分析的思路流程如下：

首先对数据进行**预处理**，排除没有意义的订单。依据附件 1，对于 240 周内所得样本订单不足的供货商，由于其数据过少，难以在忽略偶然因素的影响下确定其供货特点，故对其进行排除。该合格指标可通过订单数量、总订货量等等进行判断。

构建供货特征**评价指标**，对预备供货商集进行量化分析。此处的指标评定参考文献并结合该数据集分析进行构建。根据历史订单情况给每个供货商打分，并且引入层次分析法、熵权法等打分式模型，将各指标融合多角度判断评价。

根据该供应商评价模型，确定 50 家最具合作潜力的供应商，列表给出在文中。

#### 数据预处理
首先，根据历史 5 年供应商的供货总量，分离开供货总量过小的供应商，使得数据更有参考价值。作出图1可知，选取总供货量 100 为分界线，先行考虑大于 100 的供应商，共筛选得到 176 家供应商。同时，可以检验本题中不存在“延迟交货”这一现象，即并不存在某订单量为 0，而相应的供货量不为 0 的情况。
![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig6.png)
<div align="center">图 1: 历史总销售量供应商分布</div>

指标选择：
整个供应链系统是一种人工和自然相结合的多变量、多目标、多约束条件的复杂非线性开放系统，供应商系统是这当中的一个子系统。对于这类系统的指标设计，应遵循如下原则：

（一）指标系统性。由于系统受到多重因素影响，所以选择的参数要尽可能全面、系统地反映供
应商目前地综合水平，并对未来有一定指导。

（二）指标科学性。将绝对数指标和相对数指标相结合，通过绝对数指标反映出供应商总量上、
规模上地情况，通过相对数指标反映出速度和比率等。两类指标相辅相成，结合分析，可以更为准
确地反映实际情况。

（三）指标实用性。这主要包括指标的可计算性及计算所需数据的可行性。指标具有一定的灵活
性，使得供应商根据情况合理利用。

为了量化第 i 家供应商的供货特征，从企业合作能力、绩效、灵活性等角度定义如下指标：
![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig7.png)
<div align="center">表 1: 指标确定</div>

下面给出这 9 个指标的实际意义：

$f^{(1)}_i$供货及时性反映供应商订单的完成率，鉴于上文已经验证了不存在“延迟交货”的情况。

$f^{(2)}_i$交货准确度衡量了供货量与订货量之间的偏差情况。

$f^{(3)}_i$供货波动性反映供货与订货的差量这组数据的离散程度，体现了供货是否稳定。

$f^{(4)}_i$供货量方差体现供应商所能供货的波动性大小，这是一个极小型指标。

$f^{(5)}_i$产品占有率反映供应商在供应链中的重要性，以此反映它对于保障生产的重要性。

$f^{(6)}_i$信誉度考虑了在订货数量大小有差别时，供货商的供货数据有效性并不相同。能够很好满
足更大量的订货，比满足较小的订货，应当更有参考价值。这是一个极小型指标。

$f^{(7)}_i$平均忙期量比单纯计算供货量的平均值更有意义，它去除了供应商没有订单而导致的供
货量为 0 的情况。这个指标在第二问也有不小的作用。

$f^{(8)}_i$批量柔性用来衡量供应商的速度，如果企业增加订单，供应商在供给批量中最大能增加的
百分比。

$f^{(9)}_i$供给弹性形容的是供应商能够同时满足大订货量和小订货量的能力。

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig14.png)
<div align="center">图 2 原文指标的选择</div>

在作者引用论文中，并未给出各项指标的具体表达式，此处作者将各项表达式转化为了具体的公式表述。

#### 层次分析法+熵值法在综合评价中的应用
内容回顾

##### （一）熵值法初步确定权重
**step1.数据标准化**
设$X_{ij}$表示评价对象 i 在第 j 个指标处的取值。对于极大型指标和极小型指标，应用不同的方法对其进行标准化。对极大型指标，定义标准化后的值$Y{ij}=\frac{X{ij}-min_iX_j}{max_iX_j-min_iX_j}$;对极小型指标，定义标准化后的值$Y{ij}=\frac{max_iX_j-X{ij}}{max_iX_j-min_iX_j}$

**step2.求出各指标的信息熵**
根据信息熵的定义，一组数据的信息熵
$$
E_j=-\frac{1}{lnn}\sum^n_{i=1}p_{ij}lnp_{ij}
$$
其中$p_{ij}=\frac{Y_{ij}}{\sum_{i=1}^nY_{ij}}$，如果$p_{ij}=0$则$\lim\limits_{p_{ij}\rightarrow 0}p_{ij}lnp_{ij}=0$

差异系数$g_j=1-E_j$

**step3. 确定各指标权重**
利用step2中得到的差异系数，计算出权重
$$
W_j=\frac{g_j}{\sum_{j=1}^mg_j}
$$

**step4. 计算第i个评价对象的综合评价值**
$f_i=\sum_{j=1}^m{w_{ij}p{ij}}$
评价值越大越好

##### (二) 层次分析法适度修正
原因：由于熵权法仅是根据数据自身混乱程度，来比较得出各个指标的权重。但根据实际操作结果来看，这难免会出现与事实不吻合的情形，需要在适度范围内修正。于是加入人工干预，以熵权法得到的权重作为参考，应用层次分析法改进此评价体系。

层次分析法是一种系统性的分析方法。简洁使用，需要定量数据。
本问中将 9 个指标归类并修改为二级指标，新定义三个一级指标：**绩效评价、战略意义、潜力
估计**，得到如下图2所示的评价指标体系。二级指标相关部分的权重依据熵权法确定；一级指标相
互间的权重查阅文献，利用层次分析法得到，具体步骤如下：
![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig10.png)
<div align="center">图 3: 评价指标体系</div>

**step1. 计算比较矩阵的特征值**
将最大特征值的特征向量归一化，作为权向量。确定各因素
的权重。

**step2. 一致性检验防止确定权重随意而导致矛盾**
引入一致性指标 CI 和随机一致性指标
RI, 计算一致性比率 $\frac{CI}{RI}$ , 当小于 0.1 则满足条件。

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig15.png)
<div align="center">图 4: 多层层次分析法的一致性检验</div>

**注意：**
如果需要把这两种方法结合一起使用解决某个实际问题，那么这个实际问题应该有以下特点：

部分或全部评价指标适合使用层次结构模型进行问题的抽象；

部分评价指标满足客观赋权的计算条件（有多个样本的客观数据）。
#### 模型求解
原文所给假设：注意到指标中供货量方差$f^{(4)}_i$、供货波动性$f^{(3)}_i$ 、交货准确度$f^{(2)}_i$是极小型指标其余是极大型指标，分别用不同的方法将它们标准化，并且均化为极大型指标，便于观察。

首先，运用 SPSSAU 求出各指标权重，如下图5蓝字所示。譬如“供给弹性”指标，它在评价保障生产的重要性时作用较为次要，但因为其离散程度较大，解得的权重相应较大，与具体情况矛盾，需要进行合理范围内的修正。

然后，利用层次分析法，根据文献与经验，给出比较矩阵
$$\begin{bmatrix}
1&0.2&0.167\\
5&1&0.769\\
6&1.3&1\\
\end{bmatrix}$$
解得绩效评价、战略意义、潜力估计对应的权重为 40.467%,51.219%,8.314%, 且通过一致性检验。结合先前二级指标的权重，得到修正后的指标权重，如下图5黄字所示。
![fig11](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig11.png)
<div align="center">图 5: 层次分析法修正前后权重的比较</div>

基于修正后的权重，对供应商$ i $进行打分。记分数为 $w$, 则
$$
w_i=0.101f^{(1)}_i+0.198f^{(2)}_i+0.097f^{(3)}_i+0.008f^{(4)}_i+0.225f^{(5)}_i+0.092+f^{(6)}_i+0.195f^{(7)}_i+0.006f^{(8)}_i+0.077f^{(9)}_i
$$

#### 模型修正
注意到信誉度实际上为极小型指标，文中上下存在矛盾，且经对信誉度指标定义分析，应该属于极小型指标，所以对原模型进行一定修正，见图4。
![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig12.png)
<div align="center">图 6: 修正后的熵值法权重</div>


![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig9.png)
<div align="center">图 7: 修正后层次分析法修正前后权重的比较</div>

重新排名，见图8。（红字为从该排位起，与原文名次不同）

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig13.png)
<div align="center">图 8: 修正后排名</div>

注：在实际的比赛中，层次分析法不建议作为主模型进行使用。因为这样的模型对于权重确定有一定的开放性，如果数据来源没有明确文献支撑，会显得不够严谨。
本题中采用的基础模型为熵权法，并辅以层次分析法进行调整，采用不同模型进行糅合使用，可以给评委老师更好地印象。

### 代码实现

In [1]:
# 第一个模块——对供应及时性的考量，可以考虑称之为“履约率”
#目标：给出shape=（402，）的数列，对应各个供应商的供货及时性
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='企业的订货量（m³）')
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')

temp = df[df.columns[2:]]
temp1 = df1[df1.columns[2:]]

temp = np.array(temp)
temp1 = np.array(temp1)

bool_m=temp!=0
bool_m1=temp1!=0

bool_m_count=np.sum(bool_m,axis=1)
bool_m1_count=np.sum(bool_m1,axis=1)

result=(bool_m1_count)/(bool_m_count)
#print(result)

In [2]:
#第二个模块——产品占有率
#目标：给出三个数列，对应三种原材料各个供应商的产品占有率
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')
dfA_bool=df1["材料分类"] == "A"
dfA=df1[dfA_bool]

dfB_bool=df1["材料分类"] == "B"
dfB=df1[dfB_bool]

dfC_bool=df1["材料分类"] == "C"
dfC=df1[dfC_bool]

def every_week_total_offer(df):
    temp = df[df.columns[2:]]
    return temp.values,np.sum(temp.values,axis=0)

A_offer,A_sum=every_week_total_offer(dfA)
B_offer,B_sum=every_week_total_offer(dfB)
C_offer,C_sum=every_week_total_offer(dfC)


def rate(offer,sum,n):
    position=offer[n]!=0
    
    return np.mean(offer[n][position]/sum[position])

A=[]
B=[]
C=[]

for i in range(A_offer.shape[0]):
    A.append(rate(A_offer,A_sum,i))
for i in range(B_offer.shape[0]):
    B.append(rate(B_offer,B_sum,i))
for i in range(C_offer.shape[0]):
    C.append(rate(C_offer,C_sum,i))
    
#print(A)
#print(B)
#print(C)

In [3]:
#第三个模块——批量柔性，也即每一次最大增加百分比
#目标：给出一个数列，对应每个供应商的批量柔性指标
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='企业的订货量（m³）')
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')

temp = df[df.columns[2:]]
temp1 = df1[df1.columns[2:]]

delta=temp1-temp
data=delta/temp
data=data.fillna(0)

result=np.max(data.values,axis=1)
#print(result)

In [4]:
#第四个模块——交货准确度
#目标：给出一个数列，对应每个供应商的交货准确度指标
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='企业的订货量（m³）')
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')

temp = df[df.columns[2:]]
temp1 = df1[df1.columns[2:]]
delta=temp1-temp
data=delta/temp
data=data*data
data=data.fillna(0)
data=data.values

result=[]

for i in range(402):
    position=data[i]!=0
    result.append(np.mean(data[i][position]))

#print(result)

In [5]:
#第五个模块——供货波动性
#目标：给出一个数列，对应每个供货商波动性的指标
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='企业的订货量（m³）')
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')

temp = df[df.columns[2:]]
temp1 = df1[df1.columns[2:]]
delta=abs(temp1-temp)
'''
data=delta/temp
data=data.fillna(0)  #这几步是为了排除未产生的订单
data*=temp
data=data.values
'''
data=delta
data=data.values
result=[]

for i in range(402):
    position=data[i]!=0
    result.append(np.std(data[i][position],ddof=1)/np.mean(data[i][position]))   
#print(result)

In [6]:
#第六个模块——信誉度
#目标：给出一个数列，对应每个供货商信誉度的指标
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='企业的订货量（m³）')
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')

temp = df[df.columns[2:]]
temp1 = df1[df1.columns[2:]]
delta=temp1-temp
data=delta/temp
data=data*data*np.log(temp)
data=data.fillna(0)
data=data.values

result=[]

for i in range(402):
    position=data[i]!=0
    result.append(np.mean(data[i][position]))
    
#print(result)

E:\anaconda\lib\site-packages\ipykernel_launcher.py:14: RuntimeWarning: divide by zero encountered in log
  
E:\anaconda\lib\site-packages\numpy\core\fromnumeric.py:3441: RuntimeWarning: Mean of empty slice.
  out=out, **kwargs)
E:\anaconda\lib\site-packages\numpy\core\_methods.py:189: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)


In [7]:
#第七个模块——供给弹性
#目标：给出一个数列，对应每个供应商供给弹性的指标
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='企业的订货量（m³）')
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')

temp = df[df.columns[2:]]
temp1 = df1[df1.columns[2:]]

data=temp1/temp
data=temp*data
data=data.fillna(0)
data=data.values
result=[]

for i in range(402):
    position=data[i]!=0
    result.append(np.var(data[i][position],ddof=1)/np.mean(data[i][position]))

#print(result)

E:\anaconda\lib\site-packages\numpy\core\fromnumeric.py:3724: RuntimeWarning: Degrees of freedom <= 0 for slice
  **kwargs)
E:\anaconda\lib\site-packages\numpy\core\_methods.py:254: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)


In [19]:
#供货量方差
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='企业的订货量（m³）')
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')

temp = df[df.columns[2:]]
temp1 = df1[df1.columns[2:]]

data=temp1/temp
data=temp*data
data=data.fillna(0)
data=data.values
result=[]

for i in range(402):
    position=data[i]!=0
    result.append(np.var(data[i][position],ddof=1))

#print(result)

D:\anaconda3\lib\site-packages\numpy\core\fromnumeric.py:3723: RuntimeWarning: Degrees of freedom <= 0 for slice
  return _methods._var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,
D:\anaconda3\lib\site-packages\numpy\core\_methods.py:254: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)


In [21]:
#平均忙期量
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='企业的订货量（m³）')
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')

temp = df[df.columns[2:]]
temp1 = df1[df1.columns[2:]]

data=temp1/temp
data=temp*data
data=data.fillna(0)
data=data.values
result=[]

for i in range(402):
    position=data[i]>5
    result.append(np.mean(data[i][position]))

#print(result)


D:\anaconda3\lib\site-packages\numpy\core\fromnumeric.py:3440: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
D:\anaconda3\lib\site-packages\numpy\core\_methods.py:189: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)


In [23]:
#总供货量
import pandas as pd
import matplotlib as plt
import numpy as np
from collections import Counter
df = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='企业的订货量（m³）')
df1 = pd.read_excel("附件1 近5年402家供应商的相关数据.xlsx",sheet_name='供应商的供货量（m³）')

temp = df[df.columns[2:]]
temp1 = df1[df1.columns[2:]]

data=temp1/temp
data=temp*data
data=data.fillna(0)
data=data.values
result=[]

for i in range(402):
    position=data[i]!=0
    result.append(np.sum(data[i][position]))

#print(result)

In [1]:
#熵值法
import numpy as np
a=np.loadtxt("熵值法.txt")
l=a[:,0] #供货及时性
z=a[:,1] #产品占有率
r=a[:,2] #批量柔性
j=a[:,3] #交货准确度
d=a[:,4] #供货波动性
x=a[:,5] #信誉度
t=a[:,6] #供应弹性
f=a[:,7] #方差
m=a[:,8] #平均忙期量

def NMS(y):
    max1=np.max(y)
    min1=np.min(y)
    return (max1-y)/(max1-min1)

def MMS(y):
    max1=np.max(y)
    min1=np.min(y)
    return (y-min1)/(max1-min1)

def get_g(y):    #差异系数
    ans=0
    p= y/sum(y)
    for i in range(176):
        if p[i]>0:
            tmp=-np.log(p[i])*p[i]
            ans+=tmp
    ans=ans/np.log(176)
    return 1-ans

def final_weight(y):
    w=y/sum(y)
    return w

j=NMS(j);j=get_g(j)
x=NMS(x);x=get_g(x)
f=NMS(f);f=get_g(f)
m=MMS(m);m=get_g(m)
t=MMS(t);t=get_g(t)
l=MMS(l);l=get_g(l)
r=MMS(r);r=get_g(r)
z=MMS(z);z=get_g(z)
d=NMS(d);d=get_g(d)

b=[]
jx=[]
zl=[]
ql=[]
jx=[l,j,d,f] #绩效评价
zl=[z,x,m]  #战略意义
ql=[r,t]    #潜力估计
jx=final_weight(jx)
zl=final_weight(zl)
ql=final_weight(ql)
print("-----------------")
print("仅熵权法")
list1=["供应及时性","交货准确度","供货波动性","方差"]
list2=["市场占有率","信誉度","平均忙期量"]
list3=["批量柔性","供货弹性"]
print(dict(zip(list1,jx)))
print(dict(zip(list2,zl)))
print(dict(zip(list3,ql)))
print("-----------------")
print("熵权法+层次分析法")
jx1=0.4046*jx
zl1=0.5122*zl
ql1=0.0832*ql
list4=["供应及时性","交货准确度","供货波动性","方差"]
list5=["市场占有率","信誉度","平均忙期量"]
list6=["批量柔性","供货弹性"]
print(dict(zip(list4,jx1)))
print(dict(zip(list5,zl1)))
print(dict(zip(list6,ql1)))



-----------------
仅熵权法
{'供应及时性': 0.24331609479850239, '交货准确度': 0.4817463650205832, '供货波动性': 0.25272316624634666, '方差': 0.022214373934567763}
{'市场占有率': 0.5099618116967624, '信誉度': 0.010238589187249904, '平均忙期量': 0.4797995991159878}
{'批量柔性': 0.04423311266702876, '供货弹性': 0.9557668873329712}
-----------------
熵权法+层次分析法
{'供应及时性': 0.09844569195547406, '交货准确度': 0.19491457928732797, '供货波动性': 0.10225179306327187, '方差': 0.008987935693926117}
{'市场占有率': 0.26120243995108167, '信誉度': 0.005244205381709401, '平均忙期量': 0.24575335466720893}
{'批量柔性': 0.0036801949738967926, '供货弹性': 0.0795198050261032}


![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/fig9.png)
<div align="center">图 9: 修正后层次分析法修正前后权重的比较</div>

In [8]:
import numpy as np
class AHP:
  """
  相关信息的传入和准备
  """

  def __init__(self, array):
    ## 记录矩阵相关信息
    self.array = array
    ## 记录矩阵大小
    self.n = array.shape[0]
    # 初始化RI值，用于一致性检验
    self.RI_list = [0, 0, 0.52, 0.89, 1.12, 1.26, 1.36, 1.41, 1.46, 1.49, 1.52, 1.54, 1.56, 1.58,
            1.59]
    # 矩阵的特征值和特征向量
    self.eig_val, self.eig_vector = np.linalg.eig(self.array)
    # 矩阵的最大特征值
    self.max_eig_val = np.max(self.eig_val)
    # 矩阵最大特征值对应的特征向量
    self.max_eig_vector = self.eig_vector[:, np.argmax(self.eig_val)].real
    # 矩阵的一致性指标CI
    self.CI_val = (self.max_eig_val - self.n) / (self.n - 1)
    # 矩阵的一致性比例CR
    self.CR_val = self.CI_val / (self.RI_list[self.n - 1])

  """
  一致性判断
  """

  def test_consist(self):
    # 打印矩阵的一致性指标CI和一致性比例CR
    print("判断矩阵的CI值为：" + str(self.CI_val))
    print("判断矩阵的CR值为：" + str(self.CR_val))
    # 进行一致性检验判断
    if self.n == 2: # 当只有两个子因素的情况
      print("仅包含两个子因素，不存在一致性问题")
    else:
      if self.CR_val < 0.1: # CR值小于0.1，可以通过一致性检验
        print("判断矩阵的CR值为" + str(self.CR_val) + ",通过一致性检验")
        return True
      else: # CR值大于0.1, 一致性检验不通过
        print("判断矩阵的CR值为" + str(self.CR_val) + "未通过一致性检验")
        return False

  """
  算术平均法求权重
  """

  def cal_weight_by_arithmetic_method(self):
    # 求矩阵的每列的和
    col_sum = np.sum(self.array, axis=0)
    # 将判断矩阵按照列归一化
    array_normed = self.array / col_sum
    print("列向量归一化以后的矩阵是：")
    print(array_normed)
    # 计算权重向量
    array_weight = np.sum(array_normed, axis=1) / self.n
    # 打印权重向量
    print("算术平均法计算得到的权重向量w为：\n", array_weight)
    # 返回权重向量的值
    return array_weight

  """
  几何平均法求权重
  """

  def cal_weight__by_geometric_method(self):
    # 求矩阵的每列的积
    col_product = np.product(self.array, axis=0)
    # 将得到的积向量的每个分量进行开n次方
    array_power = np.power(col_product, 1 / self.n)
    # 将列向量归一化
    array_weight = array_power / np.sum(array_power)
    # 打印权重向量
    print("几何平均法计算得到的权重向量为：\n", array_weight)
    # 返回权重向量的值
    return array_weight

  """
  特征值法求权重
  """

  def cal_weight__by_eigenvalue_method(self):
    # 将矩阵最大特征值对应的特征向量进行归一化处理就得到了权重
    array_weight = self.max_eig_vector / np.sum(self.max_eig_vector)
    # 打印权重向量
    print("特征值法计算得到的权重向量为：\n", array_weight)
    # 返回权重向量的值
    return array_weight


if __name__ == "__main__":
  # 给出判断矩阵
  b = np.array([[1,0.2,0.167],
[5,1,0.769],
[6,1.3,1]])  #就这里每次更新成新的矩阵就行
  print("初始输入的矩阵为：\n",b)
  test = AHP(b).test_consist()
  # 算术平均法求权重
  #weight1 = AHP(b).cal_weight_by_arithmetic_method()
  # 几何平均法求权重
  #weight2 = AHP(b).cal_weight__by_geometric_method()
  # 特征值法求权重
  weight3 = AHP(b).cal_weight__by_eigenvalue_method()

初始输入的矩阵为：
 [[1.    0.2   0.167]
 [5.    1.    0.769]
 [6.    1.3   1.   ]]
判断矩阵的CI值为：(0.0006495600019311709+0j)
判断矩阵的CR值为：(0.0012491538498676364+0j)
判断矩阵的CR值为(0.0012491538498676364+0j),通过一致性检验
特征值法计算得到的权重向量为：
 [0.08317695 0.40462117 0.51220189]


## 2016年B-1题

主讲人：侯力广

### 题目回顾

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/2016B题目.png)

第一题要求选取合适的指标并建立评价体系， 用以评价小区由“ 封闭” 变为“开放”后对周边道路通行的影响。

### 评价指标体系的构建

* 评价指标选取原则
  1. 科学性：
   客观真实的反映评价对象的特征，具有典型代表性(不重不漏)
  2. 系统性：
   各指标之间有一定的逻辑关系和内在联系
  3. 可比性：
   指标选取的计算量度和计算方法一致统一
  4. 可观测性：简单明了、微观性强、便于收集
* 评价指标筛选方法
  
  目的：合理选取主要指标，剔除次要指标，简化评价体系
![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/指标筛选1.png)
![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/指标筛选2.png)

* 预处理必要性
  1. 数量级的差异将导致量级较大的属性占据主导地位；
  2. 数量级的差异将导致迭代收敛速度减慢；
  3. 依赖于样本距离的算法对于数据的数量级非常敏感。
  
  为了统一比较的标准，保证结果的可靠性，我们在分析数据之前，需要对原始变量进行一定的处理，将原始数据转化为无量纲、无数量级差异的标准化数值，消除不同指标之间因属性不同而带来的影响，从而使结果更具有可比性

* 指标一致化
  * 极大型指标(效益型指标)：指标值越大越好
  * 极小型指标(成本型指标)：指标值越小越好
  * 居中型指标：指标值适中最好
  * 区间型指标：指标值在某个区间内最好

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/指标一致化.png)

* 指标无量纲化




![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/指标无量纲化1.png)

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/指标无量纲化2.png)

### 评价模型的建立

* 论文一：模糊综合评价法

* 论文二：层次分析法

### 论文1

(1)构建评价指标体系

* 极大型：车流速度
* 极小型：车流密度，行程时间，延误时间，道路饱和度，交叉路口阻塞率

  根据指标选取原则选择六个评价指标,  查找文献得到相关数据，数据标准化处理如表格所示

(2)建立模糊评价模型

* 建立因素集

  因素集是以影响评价对象的各种因素为元素所组成的一个集合，通常用$U$表示，$U=(u_1,u_2,\cdots,u_n)$，其中$u_i$代表影响评价对象的第$i$个因素。例如综合评判员工的表现，其因素集可为：
  
  $U=${$政治表现u_1，工作能力u_2，工作态度u_3，工作成绩u_4$}。


  这里因素集$U=(u_1,u_2,u_3,u_4,u_5)$  
  $u_1$~$u_5$各对应封闭式小区、第一次增路、第二次增路、第三次增路、第四次增路

* 建立评价集

  评价集是评价者对评价对象可能做出的各种结果所组成的集合，通常用$V$表示，$V=(v_1,v_2,\cdots,v_n)$，其中元素$v_j$代表第$j$种评价结果。例如综合评判员工的表现，其评价集可为：

  $V=${$优秀v_1，良好v_2，中等v_3，较差v_4，中很差v_5$}。

  这里评价集$V=(v_1,v_2,v_3,v_4,v_5,v_6)$，
  $v_1$~$v_6$分别表示对车流密度、车流速度、行程时间、延误时间、道路饱和度、交叉路口阻塞率造成影响。

* 建立隶属矩阵



 <1>隶属度函数：



 若对论域（研究的范围）$U$中的任一元素$x$，都有一个数$A(x)\in[0，1]$与之对应，则称$A$为$U$上的模糊集，$A(x)$称为$x$对$A$的隶属度。当$x$在$U$中变动时，$A(x)$就是一个函数，称为$A$的隶属函数。隶属度$A(x)$越接近于$1$，表示$x$属于$A$的程度越高，$A(x)$越接近于$0$，表示$x$属于$A$的程度越低。



 <2>隶属度函数确定方法

 

 1.模糊统计法：找多个人去对同个模糊概念进行描述，用隶属频率去定义隶属度。(数模比赛中很少用，要设计发放问卷，可能来不及，但实际做研究用的较多)



 2.借助已有的客观尺度 (需要有合适的指标，并能收集到数据)

<table>
     <tr>
        <th>论域</th>
        <th>模糊集</th>
        <th>隶属度</th>
     </tr>
     <tr>
        <td>设备</td>
        <td>设备完好</td>
        <td>设备完好率</td>
     </tr>
     <tr>
        <td>产品</td>
        <td>质量稳定</td>
        <td>正品率</td>
     </tr>
     <tr>
        <td>家庭</td>
        <td>小康家庭</td>
        <td>恩格尔系数</td>
     </tr>
 </table>

选取的指标必须介于0,1之间(隶属度介于0,1之间)。如果不是，可以对指标归一化



 3.指派法 (根据问题的性质直接套⽤某些分布 作为⾪属函数，主观性较强)



![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/指派法隶属度.png)


 大气污染物评价标准：

<table>
     <tr>
        <th> </th>
        <th>I级</th>
        <th>II级</th>
        <th>III级</th>
        <th>IV级</th>
     </tr>
     <tr>
        <th>污染物1</th>
        <th>0.05</th>
        <th>0.15</th>
        <th>0.25</th>
        <th>0.50</th>
     </tr>
     <tr>
        <th>污染物2</th>
        <th>0.12</th>
        <th>0.30</th>
        <th>0.50</th>
        <th>1.00</th>
     </tr>
     <tr>
        <th>污染物3</th>
        <th>0.10</th>
        <th>0.10</th>
        <th>0.15</th>
        <th>0.30</th>
      </tr>
      <tr>
          <th>污染物4</th>
        <th>4.00</th>
        <th>4.00</th>
        <th>6.00</th>
        <th>10.00</th>
     </tr>
     <tr>
        <th>污染物5</th>
        <th>0.05</th>
        <th>0.15</th>
        <th>0.25</th>
        <th>0.50</th>
     </tr>
     <tr>
        <th>污染物6</th>
        <th>0.12</th>
        <th>0.16</th>
        <th>0.20</th>
        <th>0.40</th>
     </tr>
 </table>

 例：确定污染物i对空气质量各等级的隶属函数



![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/隶属度例子1.png)



 <3>建立隶属矩阵



 因素集$U$中第$i$个元素对评价集$V$中第$j$个元素的隶属度为$m_{ij}$,可得所求$M_{i*j}$

 例：某日测得污染物的日均浓度为$(0.07,0.20,0.123,5.00,0.08,0.14)$，则污染物i对空气质量各等级的隶属度矩阵为：
 
$$
R=\begin{bmatrix}
    0.8&0.2&0&0\\
  0.56&0.44&0&0\\
   0&0.6&0.4&0\\
    0&0.5&0.5&0\\
    0.7&0.3&0&0\\
    0.5&0.5&0&0
    \end{bmatrix}
$$


本文直接给出选择的隶属函数，带入各指标数据，得到各小区类型对各项指标的隶属度矩阵M。(matlab)

* 确定各因素权重并求解评价向量

  通常做法 ：
  1. 确定各影响因素$u_i$的权重$a_i$，得到权重向量$A=(a_1,\cdots,a_5)$
  
     没有数据时可用层次分析法，有数据时可用熵权法。
  2. 对隶属度矩阵的每一行做标准化，得到新的评价矩阵$M'_{5*6}$
  3. 计算评价向量$B_{1*6}=A_{1*5}*M'_{5*6}$(矩阵乘法)，得出各指标的综合权重

  本文做法：
  1. 熵权法求解各评价指标$v_i$的权重，得到权重向量
  
      $v=(0.0932,0.0381,0.2488,0.3827,0.0383,0.1989)_{1*6}$

  2. 对隶属度矩阵的每一列做标准化，得到新的评价矩阵$M'_{5*6}$
  3. 计算评价向量$B_{5*1}=M'_{5*6}*v_{6*1}$，得到各因素的综合权重（matlab）

  个人修改：
  
  * 熵权法原理：

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/熵权法确定权重.png)

  * 评价向量$B=(0.1766,0.1809,0.1698,0.1316,0.1871,0.1541)$

```%论文1-熵权值求权重
A=[58 38 14 8 57 10
50 45 11 9 52 12
42 47 8 12 50 15
45 42 12 15 46 16
47 44 13 10 49 13];%统计数据A 5x6

[m,n]=size(A);
h =max(A);%每一列最大值
H=repmat(h,m,1);%向量h排列mx1的矩阵
Mij=A./H%隶属度矩阵

A2=sum(Mij)%Mij矩阵每一列求和
A22=repmat(A2,m,1);%向量A2排列mx1的矩阵
Qij=Mij./A22%隶属度矩阵归一化
%ss=sum(Qij)%检验归一化

Fj=-1/log(5)*Qij.*log(Qij);
fj=sum(Fj);%熵值
dj=1-fj;%差异系数
v=dj/sum(dj)%权重系数
B=Qij*v';%最终评价
disp(B)%显示结果
```

这一段是论文1中熵权值求权重的matlab代码。

#### 论文2

(1)构建评价指标体系
* R型聚类法

  用相关系数表示变量相似度，得到相关系数矩阵：

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/相似度矩阵.png)

  用最长距离法解决变量聚类问题(matlab)

  ```matlab
  a=[0.85 1.00 0.83 0.90 0.85 0.30 0.10 0.32 0.21;
  0.90 0.83 1.00 0.79 0.92 0.21 0.11 0.16 0.26;
  0.81 0.90 0.79 1.00 0.70 0.30 0.19 0.14 0.27;
  0.76 0.85 0.92 0.70 1.000 0.17 0.10 0.12 0.25;
  0.30 0.30 0.21 0.30 0.17 1.00 0.92 0.23 0.17;
  0.42 0.10 0.11 0.19 0.10 0.92 1.00 0.25 0.14;
  0.21 0.32 0.16 0.14 0.12 0.23 0.25 1.00 0.95;
  0.33 0.21 0.26 0.27 0.25 0.17 0.14 0.95 1.00];
  d=1-abs(a);%进行数据变换，把相关系数转化为距离
  %y=linkage(d,'complete');%按最大距离法聚类
  y=linkage(d,'average');%按平均距离法聚类
  j=dendrogram(y);%画聚类图
  L=cluster(y,'maxclust',3);%把变量划分成3类
  for i=1:3
  b=find(L==i);
  b=reshape(b,1,length(b));%将b重构为1xlength(b)的数组
  fprintf('第%d 类的有%s\n',i,int2str(b));%intstr(b)将b转换为整数
  end
  ```

* 聚类结果：

1. 影响主路通行能力的因素：
   
   车道数，道路面积率，拥堵系数，路旁干扰系数，区内支路饱和度
2. 影响安全性的因素：
   
   交叉路口个数，车辆种类
3. 影响便捷度的因素：
   
   可达度，抗堵塞能力。

(2)建立评价模型

* 建立层次结构模型

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/聚类结果.png)

* 建立判别矩阵

按下表给出的9个重要性等级尽行赋值，按两两比较结果构成判别矩阵($a_{ij}a_{ji}=1$)。

![](https://cdn.jsdelivr.net/gh/arctanln2/PictureStorage@master/blogs/pictures/AHP比例.png)


例：$C_1$比 $C_2$稍微重要

* 求解层间权重及一致性检验

由于判断矩阵$A$为正矩阵，由佩罗(Perron)定理可知：正互反矩阵一定存在一个最大特征值$\lambda_{max}$，且$\lambda_{max}$所对应的特征向量$x$为正向量，即$Ax=\lambda_{max}x$。将$x$归一化作为权向量$w$，即$Aw=\lambda_{max}w$

 引入一致性指标$CI$和随机一致性指标$RI$，计算一致性比率$CR$
$$
CI=\frac{\lambda_{max}-n}{n-1},\;CR=\frac{CI}{RI}
$$


 $CR<0.1$,则认为判断矩阵具有满意的一致性，否则就要调整判断矩阵直至$CR<0.1$

```matlab
%层次分析法判别矩阵的一致性检验代码
clc,clear
%P=[1 4 5;1/4 1 3;1/5 1/3 1];%判别矩阵
P=[1.0000 3.0000 4.0000;
0.3333 1.0000 2.0000;
0.2500 0.5000 1.0000];%判别矩阵
[L,G]=eig(P);%求特征根和相应的特征向量
w=L(:,1)/sum(L(:,1));%归一化特征向量近似为权重值
max=max(eig(P));%求最大特征根
n=size(P);
CI=(max-3)/(3-1);
RI=1.12; %查表对应 n=3 的情况
CR=CI/RI
if CR>=0.1
error('A 不通过一致性检验')
end
```

这是层次分析法判别矩阵的一致性检验代码。

* 计算总权重

Remark: 对于指标的选取，要尽可能选择公共的指标；另外在赛前要储备一些可以找数据的网站以备需要，要提升自己找数据的能力。